### PR TITLE
[DSCP-82] Remove hardcoded dependency on binary from avl package

### PR DESCRIPTION
**Pull request by Heimdell**
Removed binary instances, put `Serialisable` interface instead.

### CODE REVIEW

File: src/Data/Tree/AVL/Deletion.hs [hunk 1]
```diff
@@ -1,8 +1,9 @@
 
-{-# LANGUAGE MultiWayIf     #-}
-{-# LANGUAGE NamedFieldPuns #-}
-{-# LANGUAGE LambdaCase #-}
-{-# LANGUAGE ExplicitForAll #-}
+{-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
```
Comment (volhovm) [inline src/Data/Tree/AVL/Deletion.hs]: LambdaCase should go into the global config (cabal).
https://github.com/serokell/serokell-util/blob/master/serokell-style.md#default-extensions

I'd also put  `FlexibleContexts` there as well. 

This is just a nit though.

File: src/Data/Tree/AVL/HashMapStore.hs [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
-{-# LANGUAGE FlexibleInstances #-}
-{-# LANGUAGE FlexibleInstances #-}
-{-# LANGUAGE MultiParamTypeClasses  #-}
-{-# LANGUAGE FlexibleContexts       #-}
-{-# LANGUAGE ScopedTypeVariables       #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
```
Comment (volhovm) [inline src/Data/Tree/AVL/HashMapStore.hs]: ScopedTypeVariables should be default also.

File: src/Data/Tree/AVL/HashMapStore.hs [hunk 3]
```diff
@@ -21,43 +20,43 @@ import Data.Tree.AVL.Internal
...
-        modify $ insert k (encode v)
+        modify $ insert k (serialise v)
 
-instance (Show a, Typeable a) => KVStoreMonad a NullStore where
+instance (Show k, Typeable k, Serialisable k) => KVStoreMonad k NullStore where
     retrieve k = throwM (NotFound k)
     store _ _  = return ()
 
-runWithCache :: KVStoreMonad h m => Storage h -> HashMapStore h m a -> m (a, Storage h)
+runWithCache :: forall k m a. KVStoreMonad k m => Storage k -> HashMapStore k m a -> m (a, Storage k)
```
Comment (volhovm) [inline src/Data/Tree/AVL/HashMapStore.hs]: Why did you add explicit foralls here?

File: src/Data/Tree/AVL/Internal.hs [hunk 4]
```diff
@@ -82,20 +82,23 @@ data MapLayer h k v self
...
 
-deriving instance                                      Generic (Free t a)
-deriving instance (Binary (t (Free t a)), Binary a) => Binary  (Free t a)
+deriving instance Generic (Free t a)
 
 makeLenses ''MapLayer
 
+class Serialisable a where
+    serialise   :: a -> ByteString
+    deserialise :: ByteString -> Maybe a
```
Comment (volhovm) [inline src/Data/Tree/AVL/Internal.hs]: Please use `Either t` for some readable `t`. It will definitely help in the future to debug problems related to different app versions (or testnet/mainnet/cluster magic mismatch). A short class description (as a good practice, I know it's obvious what this class does) would be also great.

File: test/Common.hs [hunk 5]
```diff
@@ -38,6 +40,14 @@ import Test.QuickCheck.Instances            as T ()
 
 import qualified Data.Tree.AVL as AVL
 
+instance Binary x => AVL.Serialisable x where
```
Comment (volhovm) [inline test/Common.hs]: So the PR doesn't remove `binary` dependency and it moves to its own serialisable class only partially. What's the deal?

```
ξ> git grep Binary
test/Common.hs:import Data.Binary                               (Binary, encode, decodeOrFail)
test/Common.hs:instance Binary x => AVL.Serialisable x where
test/Common.hs:instance Binary Layer
test/Common.hs:instance Binary   AVL.Tilt
test/Common.hs:    deriving (Eq, Ord, Generic, Binary)
````

**Review by volhovm — CHANGES_REQUESTED** [review]
_please_ don't use branch names that long, it's not necessary but it takes a lot of space in console and editor :)

**Review by flyingleafe — APPROVED** [review]
I agree with @volhovm 's comments. Otherwise it's alright.
